smol/capp/generate/groundstate/solver.py

```diff
@@ -0,0 +1,553 @@
+        """
+        if (
+            self._ground_state_solution is not None
+            and self._ground_state_energy is not None
+        ):
+            warn("Ground state already solved before. Overwriting previous result!")
+        self.problem.solve(
+            solver=self.solver, warm_start=self.warm_start, **self.solver_options
+        )
+        if (
+            self.variables.value is not None
+            and self.objective_function.value is not None
+        ):
+            self._ground_state_solution = self.variables.value.astype(int)
+            self._ground_state_energy = self.objective_function.value
+        else:
+            warn(
+                "Ground state could not be solved! Try using another solver, or"
+                " fine tune solver options!"
+            )
+            self._ground_state_solution = None
+            self._ground_state_energy = None
+        return self._ground_state_solution, self._ground_state_energy
+
+    def reset(self):
+        """Clear previous solutions and reinitialize the problem."""
+        self._initialize_problem()
+
+    def _raise_unsolved(self):
+        if self._ground_state_solution is None or self._ground_state_energy is None:
+            raise RuntimeError(
+                "The ground state of the current system is not"
+                " solved before or was not successful. Call"
+                " self.solve()!"
+            )
+
+    @property
+    def ground_state_solution(self):
+        """Boolean variable values at the ground state.
+
+        Returns:
+            np.ndarray: the solution array in 0 and 1.
+        """
+        self._raise_unsolved()
+        return self._ground_state_solution
+
+    @property
+    def ground_state_energy(self):
+        """Energy per super-cell in the ground-state.
+
+        Returns:
+            float.
+        """
+        self._raise_unsolved()
+        return self._ground_state_energy
+
+    @property
+    def ground_state_occupancy(self):
+        """Encoded occupancy string of the ground-state.
+
+        Returns:
+            np.ndarray
+        """
+        self._raise_unsolved()
+        if self._ground_state_occupancy is None:
+            self._ground_state_occupancy = get_occupancy_from_variables(
+                self.sublattices,
+                self.ground_state_solution,
+                self.variable_indices,
+            )
+        return self._ground_state_occupancy
+
+    @property
+    def ground_state_structure(self):
+        """The ground state structure.
+
+        Returns:
+            Structure.
+        """
+        self._raise_unsolved()
+        if self._ground_state_structure is None:
+            self._ground_state_structure = (
+                self._ensemble.processor.structure_from_occupancy(
+                    self.ground_state_occupancy
+                )
+            )
+        return self._ground_state_structure
+
+    def as_dict(self):
```

Comment:
I'm not sure why we should have this be MSONable. Does loading from json make a noticeable time difference that simply creating a new solver instance?

Reply:
By making the solver object MSONable, I did not intend to improve efficiency, but simply wanted the users to be able to save and recheck their problem setups.

I can remove the MSONable feature if you have any ideas.

Reply:
Seems that MSONable is not very necessary. Considering that Sampler or Kernel does not need to be  MSONable, I will remove MSONization here. The users will be responsible for remembering their parameter choices and for recording their results.